Add proposal + PoC for Observatorium Go-based configuration (#507)
observatorium/observatorium: docs/proposals/20230130-go-configuration.md

## Pitfalls of Current Solution

Currently, we primarily use [jsonnet](https://jsonnet.org/) in Observatorium, to compose a "meta-project" library that allows generating configuration for a multi-tenant scalable observability system.

While jsonnet is a commonly used configuration templating language, there are certain issues with how jsonnet acts as a front-level configuration interface, which causes it to not be preferable to end-users.

Templating via jsonnet, while achievable, often results in complex, unreadable code, and due to no doc generation and low IDE support becomes virtually undiscoverable. Codifying and abstracting operational practices and knowledge in jsonnet is difficult and time-consuming, due to the lack of type safety and testing practices. In certain cases, pre-existing upstreams also cannot be used and jsonnet templating needs to be written manually.

> Furthermore, for a project like Observatorium, this leads to a complex-to-adopt and messy installation model. An end-user is forced to go through one of the below,

Reviewer: Or what if the end-user already uses jsonnet in their environment due to wider community adoption (i.e. from Grafana, Prometheus, etc)?

Author: That's fine! 

It still means the end-user would have to learn how the Observatorium jsonnet lib is constructed and then spend time navigating and modifying tricky jsonnet to customize it for their pipelines. Maybe spend even more time trying to work observatorium into their jsonnet.

With this, they can have separate easier-to-operate Go code which they can customize to output YAML they need.

In fact, this is what we do downstream with our Observatorium RBAC. We maintain it in Go because it's easier to modify than jsonnet templating: https://github.com/rhobs/configuration/blob/main/configuration/observatorium/rbac.go. 🙂

Reviewer: What I mean is that the proposal seems biased toward people that do not already know jsonnet. So saying Observatorium is complex to adopt and has a messy installation model is not a fact, it's relatively to who does it. 

This part of the proposal could be probably rewritten something in the lines of "we assume (and assumptions could be wrong) that there aren't many people out there with enough jsonnet knowledge and for people new to it its usage is complex and messy".

Author: I mean, we are jsonnet users, we deal with it on a daily basis, and know how to use it. And still, find it difficult and tricky to maintain, adopt and scale with it! I'm not sure if this is a biased view?

Reviewer 2: I think even wider community does not adopt jsonnet as fast as we envisioned.

So I assume @douglascamata you just want to rephrase existing to something like 
```suggestion
Furthermore, for a project like Observatorium, this leads to a complex-to-adopt installation model, especially for those not familiar with jsonnet. An end-user is forced to go through one of the below,
```

* Scale the steep learning curve of jsonnet initially (if needed), and understand complex Observatorium jsonnet library internals and abstractions (without documentation or discoverability) to be able to configure it exactly according to needs (sometimes even rewrite for customizability)
* Copy over pre-generated YAML in observatorium/observatorium and change it manually
* Reference generated YAML and rewrite it in templating the end-user is comfortable with

## Goals

Compose Observatorium configuration as a Go-based library using configuration-as-code concepts, that allows end-users to,
* Easily discover configuration abstractions, without having to go through steep configuration templating language learning curves and having to explore templating code in-depth
* Easily utilize pre-defined abstractions for their specific environments, and be able to override configuration at will, without hassle

> * Be able to rely on tested configuration generation, battle-tested versioning via Go modules and automatic documentation via [goDoc](https://blog.golang.org/godoc-documenting-go-code)

Reviewer 2: I fully agree. But still, having familiar and well-maintained version manager is a nice thing. Not sure if a strong goal, though. 

I am missing good IDE support with autocompletion as a goal. 🤔

> * Be able to rely on tested configuration generation, battle-tested versioning via Go modules and automatic documentation via [goDoc](https://blog.golang.org/godoc-documenting-go-code)

Author: Will add the goal to the proposal

* Be able to pull in, their own upstreams directly by just importing Go struct for their own configuration (sidecars/config files, etc)
* Have a consistent installation and usage model for Observatorium, as a customizable simple static configuration generation solution, with sane defaults

## Non-Goals

* Deprecating/removing jsonnet entirely. This is only proposing Go-based configuration as an alternative method of usage, but does not take away from exisiting jsonnet templating in any way
* Operator or operator generation. This proposal only aims for static configuration generation, and is not related to applying this configuration or automating it

## Audience of this proposal

> Observatorium Devs and Users.

Reviewer: Users are part of the audience but the proposal doesn't outline what's the user's point of view. For example, if I deploy Observatorium today using the jsonnet libs & apply some customizations there I only need to: create a jsonnetfile, add the dependency, run the jsonnet bundler tools (`jb`), and then use the `jsonnet` cli to render them. 

What would be the steps to reuse configuration with this Go system?

Author: I don't think the steps for jsonnet are quite clear, here. There are a few more steps to using jsonnet: create a jsonnetfile, add the dependency, run the jsonnet bundler tools (jb), understand how Observatorium's jsonnet lib works and is constructed (without docs/IDE support) and hunt through generated YAMLs to figure out how end state is affected, customize by writing complex jsonnet for diff envs to get to state that fits user deployment pipeline, use the jsonnet cli to render them to JSON and then use a tool like gojsontoyaml to convert to YAML.

The path for Go-based one is: 
Have Go tools installed, create a Go module and import obs/obs, check an example to figure out observatorium config abstractions or just go to pkg.go.dev for godoc or just click through + autocomplete in IDE to understand abstractions, use said abstractions and customize struct for diff envs to fit user deploy pipeline, `go run` to get YAML you need

In both cases above, we don't consider steps about learning Go or jsonnet.

Reviewer: The point of my questions was more to think about these steps and outline them in the proposal, food for thought. Particularly I am missing the point of view of the user of Observatorium in there. 

i.e. why would I have to understand Observatorium's jsonnet lib? Aren't the default values good enough there? Why in the Go based config I can simply "check examples and figure out" and not in jsonnet? 

Better outlining these things will make the proposal more attractive, imho.

Author: I think this is covered by the "Pros" section of the proposal. All of those point out what can be done in Go and not jsonnet or which one is easier. If this feels inadequate, would adding a few examples of this help here?

				"http://observatorium-xyz-rules-objstore.observatorium.svc.cluster.local:8080",
			),
			api.WithLogs(
				"http://observatorium-xyz-loki-query-frontend-http.observatorium.svc.cluster.local:3100",
				"http://observatorium-xyz-loki-distributor-http.observatorium.svc.cluster.local:3100",
				"http://observatorium-xyz-loki-ruler-http.observatorium.svc.cluster.local:3100",
				"http://observatorium-xyz-loki-querier-http.observatorium.svc.cluster.local:3100",
			),
			api.WithTraces(
				"http://observatorium-xyz-jaeger-query.observatorium.svc.cluster.local:16686/",
				"observatorium-xyz-otel-collector:4317",
			),
			api.WithRateLimiter("observatorium-xyz-gubernator.observatorium.svc.cluster.local:8081"),
			api.WithRBACYAML(string(rbacData)),
			api.WithTenantsSecret(map[string]string{"tenants.yaml": string(tenantsData)}),

			api.WithImage("quay.io/observatorium/api", "main-2023-01-24-v0.1.2-318-g5f4fdf4"),
			api.WithName("observatorium-xyz"),
			api.WithNamespace("observatorium"),
			api.WithReplicas(3),
			api.WithAPIResources(apiResources),
			api.WithServiceMonitor(),
		).Manifests(),
		"config",
	)
```

### Future Installation Model

> We can even leverage [`obsctl`](https://github.com/observatorium/obsctl) in the future to allow generating boilerplate Go code that imports Observatorium Go config and provides the user with commands to generate sane default YAML config.

Reviewer: Isn't `obsctl` becoming a tool that does many very different things? Is this okay for the project's direction?

Author: `obsctl` is a tool for making observatorium easier to interact with IMO. This just generates a boilerplate, and doesn't do anything special, customization is still a manual operation for user. However I don't feel too strongly about this, can remove this idea if others agree. 🙂

Reviewer: Totally get the objective of `obsctl`. But it's too generic and going against "do one thing and do it well". Which can be a totally fine direction, so if that's the case, all good to me.

Reviewer 2: I understand your concern @douglascamata, but I wouldn't say it's too much. obsctl still interacts with Observatorium resources in this case.
